{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 2] No such file or directory: 'python_project/'\n",
      "/home/thodoris/Desktop/python_project\n"
     ]
    }
   ],
   "source": [
    "%cd python_project/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/thodoris/Desktop/python_project\r\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/thodoris/anaconda3/lib/python3.6/site-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning\n",
      "  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import requests\n",
    "from fuzzywuzzy import fuzz\n",
    "from fuzzywuzzy import process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('chromosome 1 and position', 84),\n",
       " ('chromosome X, position 1234567', 69),\n",
       " ('chromosome 45, position 1654651', 68),\n",
       " ('chromosome 22, position 16986845', 67),\n",
       " ('chr1:1234567A>G', 30)]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "choices = [\"chromosome 45, position 1654651\", \"chromosome 22, position 16986845\", \"chromosome 1 and position\", \"chromosome X, position 1234567\", 'chr1:1234567A>G']\n",
    "process.extract(\"chomosome and , porisiton\", choices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "((0, 'Which gene is upstream to gene <gene> ?'), 95)\n",
      "Your question is closest to this one: Which gene is upstream to gene <gene> ?\n"
     ]
    }
   ],
   "source": [
    "a='Which gene is upstream to gene ?'\n",
    "def choices(x):\n",
    "    question_list=[\n",
    "                   (0,'Which gene is upstream to gene <gene> ?'),\n",
    "                   (0,'upstream to gene <gene> ?'),\n",
    "                   (1,'How many transcripts does the gene <gene> have?'), \n",
    "                   (1,'How many transcripts of the gene <gene> ?'),\n",
    "                   (2,'What is in <position> ?'),\n",
    "                   (3,'What is the name of the gene <gene> ?'),\n",
    "                   (3,'name of the gene <gene> ?'),\n",
    "                   (3,'the name of the gene <gene> ?'),\n",
    "                   (4,'Where is the location of <gene/transcript/mutation> ?'),\n",
    "                   (4,'location of <gene/transcript/mutation> ?'),\n",
    "                   (4,'Where is the <gene/transcript/mutation> ?'),\n",
    "                   (5,'Which are the refseq transcripts of the gene <gene> ?'), \n",
    "                   (5,'refseq transcripts of the gene <gene> ?'),\n",
    "                   (6,'How many exons does <gene/transcript> have ?'), \n",
    "                   (6,'How many exons of <gene/transcript> ?'),\n",
    "                   (7,'Where is the location of the CDS of the <gene/transcript> ?'),\n",
    "                   (7,'location of the CDS of the <gene/transcript> ?'),\n",
    "                   (8,'What is the primary transcript of the gene <gene> ?'), \n",
    "                   (8,'primary transcript of the gene <gene> ?'),\n",
    "                   (9,'In which gene does <transcript/position/mutation> belong ?'),\n",
    "                   (9,'In which gene is <transcript/position/mutation> ?'),\n",
    "                   (10,'In which pathways is the gene <gene> involved ?'), \n",
    "                   (10,'pathways of the gene <gene> involved ?'),\n",
    "                   (11,'Which proteins are transcribed from the <gene/transcript> ?'), \n",
    "                   (11,'transcribed proteinsof the <gene/transcript> ?'),\n",
    "                   (12,'What is the function of the gene <gene> ?'),\n",
    "                   (12,'function of the gene <gene> ?'),\n",
    "                   (13,'Which organisms have a homologous gene to <gene> ?'),\n",
    "                   (14, 'What type is the gene <gene> ?'), \n",
    "                   (14,'type of gene <gene> ?'),\n",
    "                   (15, 'What is the sequence in chromosome <> between start(int) and end(int)?'),\n",
    "                   (15,'sequence in chromosome <> between start(int) and end(int)?'),\n",
    "                   (16,'What is the allele frequency of <mutation> ?'), \n",
    "                   (17,'With which conditions is the mutation <mutation> associated ?'),\n",
    "                   (18, 'What is the rs-id of <mutation> ?'), \n",
    "                   (18,'rs-id of <mutation> '),\n",
    "                   (19,'What is the effect of <mutation> ?'),\n",
    "                   (19,'effect of <mutation>'),\n",
    "                  ]\n",
    "    for question in question_list:\n",
    "        limit=process.extractOne(x, question_list)\n",
    "    print(limit)\n",
    "    print(f'Your question is closest to this one: {limit[0][1]}')\n",
    "     #     f'{limit[2]}' +2*'\\n'+\n",
    "      #    'If you mean something else, specify your question.')\n",
    "choices(a)\n",
    "#print(list(enumerate(question_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('What is the function of the gene <gene> ?', 95)\n",
      "Your question is closest to this one: What is the function of the gene <gene> ?\n"
     ]
    }
   ],
   "source": [
    "a='What is the function of <gene> ?'\n",
    "def choices(x):\n",
    "    question_list=[\n",
    "                   'Which gene is upstream to gene <gene> ?',\n",
    "                   'upstream to gene <gene> ?',\n",
    "                   'How many transcripts does the gene <gene> have?', \n",
    "                   'How many transcripts of the gene <gene> ?',\n",
    "                   'What is in <position> ?',\n",
    "                   'What is the name of the gene <gene> ?',\n",
    "                   'name of the gene <gene> ?',\n",
    "                   'the name of the gene <gene> ?',\n",
    "                   'Where is the location of <gene/transcript/mutation> ?',\n",
    "                   'location of <gene/transcript/mutation> ?',\n",
    "                   'Where is the <gene/transcript/mutation> ?',\n",
    "                   'Which are the refseq transcripts of the gene <gene> ?', \n",
    "                   'refseq transcripts of the gene <gene> ?',\n",
    "                   'How many exons does <gene/transcript> have ?', \n",
    "                   'How many exons of <gene/transcript> ?',\n",
    "                   'Where is the location of the CDS of the <gene/transcript> ?',\n",
    "                   'location of the CDS of the <gene/transcript> ?',\n",
    "                   'What is the primary transcript of the gene <gene> ?', \n",
    "                   'primary transcript of the gene <gene> ?',\n",
    "                   'In which gene does <transcript/position/mutation> belong ?',\n",
    "                   'In which gene is <transcript/position/mutation> ?',\n",
    "                   'In which pathways is the gene <gene> involved ?', \n",
    "                   'pathways of the gene <gene> involved ?',\n",
    "                   'Which proteins are transcribed from the <gene/transcript> ?', \n",
    "                   'transcribed proteinsof the <gene/transcript> ?',\n",
    "                   'What is the function of the gene <gene> ?',\n",
    "                   'function of the gene <gene> ?',\n",
    "                   'Which organisms have a homologous gene to <gene> ?',\n",
    "                   'What type is the gene <gene> ?', \n",
    "                   'type of gene <gene> ?',\n",
    "                   'What is the sequence in chromosome <> between start(int) and end(int)?',\n",
    "                   'sequence in chromosome <> between start(int) and end(int)?',\n",
    "                   'What is the allele frequency of <mutation> ?', \n",
    "                   'With which conditions is the mutation <mutation> associated ?',\n",
    "                   'What is the rs-id of <mutation> ?', \n",
    "                   'rs-id of <mutation> ',\n",
    "                   'What is the effect of <mutation> ?',\n",
    "                   'effect of <mutation>',\n",
    "                  ]\n",
    "    for question in question_list:\n",
    "        limit=process.extractOne(x, question_list)\n",
    "    print(limit)\n",
    "    print(f'Your question is closest to this one: {limit[0]}')\n",
    "     #     f'{limit[2]}' +2*'\\n'+\n",
    "      #    'If you mean something else, specify your question.')\n",
    "choices(a)\n",
    "#print(list(enumerate(question_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene', 'ENSTSC-255']"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a='ENSTSC-255'\n",
    "def recognition(x):\n",
    "\n",
    "    #choices = []\n",
    "    #limit=process.extract(a, choices)\n",
    "\n",
    "    if x is not None:\n",
    "        \n",
    "        if 'NM_' in x and not '>' in x:\n",
    "            return ['Refseq_id', x]\n",
    "        elif 'ENST00' in x and not '>' in x:\n",
    "            return ['Ensembl_id',x]\n",
    "        elif 'rs' in x:\n",
    "            return ['Mutation rs-id', x]\n",
    "        elif x.startswith(('chr','Chr', 'cHr','chR', 'CHR', 'cHR', 'ChR')):\n",
    "            if '>' in x:\n",
    "                return ['Mutation', x]\n",
    "            else:\n",
    "                return ['Position', x]\n",
    "        elif x[0].isdigit():\n",
    "            if '>' in x:\n",
    "                return ['Mutation', x]\n",
    "            else:\n",
    "                return ['Position', x]\n",
    "        elif 'c.' in x:\n",
    "            if x.startswith(('ENST', 'NM')):\n",
    "                return ['Mutation with transcript prefix', x]\n",
    "            else:\n",
    "                return ['Mutation with gene prefix', x]\n",
    "        else:\n",
    "            return ['Gene' , x]\n",
    "    else:\n",
    "        raise Exception ('Please specify a question')\n",
    "\n",
    "    \n",
    "recognition(a)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 250867 entries, 25 to 2739372\n",
      "Data columns (total 9 columns):\n",
      "seqname      250867 non-null object\n",
      "source       250867 non-null object\n",
      "feature      250867 non-null object\n",
      "start        250867 non-null int64\n",
      "end          250867 non-null int64\n",
      "score        250867 non-null object\n",
      "strand       250867 non-null object\n",
      "frame        250867 non-null object\n",
      "attribute    250867 non-null object\n",
      "dtypes: int64(2), object(7)\n",
      "memory usage: 19.1+ MB\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "gencode = pd.read_table(\"gencode.v29.annotation.gff3\", comment=\"#\", sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start' , 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "gencode=gencode[gencode.source != 'HAVANA']\n",
    "gencode.head() \n",
    "gencode.info()\n",
    "gencodeexin=gencode.drop(['source','score','strand','frame'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript(x):\n",
    "    z  = dict([y.split('=') for y in x.split(';')])\n",
    "    return z.get('transcript_id')\n",
    "\n",
    "def get_tag(x):\n",
    "    z  = dict([y.split('=') for y in x.split(';')])\n",
    "    return z.get('tag')\n",
    "\n",
    "def get_genename(x):\n",
    "    z  = dict([y.split('=') for y in x.split(';')])\n",
    "    return z.get('gene_name')\n",
    "\n",
    "def get_geneid(x):\n",
    "    z  = dict([y.split('=') for y in x.split(';')])\n",
    "    return z.get('gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencodeexin['transcript'] = gencodeexin.apply(lambda x: get_transcript(x['attribute'])  , axis=1)\n",
    "gencodeexin['tag'] = gencodeexin.apply(lambda x: get_tag(x['attribute'])  , axis=1)\n",
    "gencodeexin['geneid'] = gencodeexin.apply(lambda x: get_geneid(x['attribute'])  , axis=1)\n",
    "gencodeexin['genename'] = gencode.apply(lambda x: get_genename(x['attribute'])  , axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript</th>\n",
       "      <th>tag</th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>69055</td>\n",
       "      <td>70108</td>\n",
       "      <td>ID=ENST00000335137.4;Parent=ENSG00000186092.6;...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>69055</td>\n",
       "      <td>70108</td>\n",
       "      <td>ID=exon:ENST00000335137.4:1;Parent=ENST0000033...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>ID=CDS:ENST00000335137.4;Parent=ENST0000033513...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chr1</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>69091</td>\n",
       "      <td>69093</td>\n",
       "      <td>ID=start_codon:ENST00000335137.4;Parent=ENST00...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>ID=stop_codon:ENST00000335137.4;Parent=ENST000...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chr1</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>69055</td>\n",
       "      <td>69090</td>\n",
       "      <td>ID=UTR5:ENST00000335137.4;Parent=ENST000003351...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr1</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>70009</td>\n",
       "      <td>70108</td>\n",
       "      <td>ID=UTR3:ENST00000335137.4;Parent=ENST000003351...</td>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname          feature  start    end  \\\n",
       "69    chr1       transcript  69055  70108   \n",
       "70    chr1             exon  69055  70108   \n",
       "71    chr1              CDS  69091  70008   \n",
       "72    chr1      start_codon  69091  69093   \n",
       "73    chr1       stop_codon  70006  70008   \n",
       "74    chr1   five_prime_UTR  69055  69090   \n",
       "75    chr1  three_prime_UTR  70009  70108   \n",
       "\n",
       "                                            attribute         transcript  \\\n",
       "69  ID=ENST00000335137.4;Parent=ENSG00000186092.6;...  ENST00000335137.4   \n",
       "70  ID=exon:ENST00000335137.4:1;Parent=ENST0000033...  ENST00000335137.4   \n",
       "71  ID=CDS:ENST00000335137.4;Parent=ENST0000033513...  ENST00000335137.4   \n",
       "72  ID=start_codon:ENST00000335137.4;Parent=ENST00...  ENST00000335137.4   \n",
       "73  ID=stop_codon:ENST00000335137.4;Parent=ENST000...  ENST00000335137.4   \n",
       "74  ID=UTR5:ENST00000335137.4;Parent=ENST000003351...  ENST00000335137.4   \n",
       "75  ID=UTR3:ENST00000335137.4;Parent=ENST000003351...  ENST00000335137.4   \n",
       "\n",
       "                              tag             geneid genename  \n",
       "69  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  \n",
       "70  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  \n",
       "71  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  \n",
       "72  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  \n",
       "73  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  \n",
       "74  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  \n",
       "75  basic,appris_principal_1,CCDS  ENSG00000186092.6    OR4F5  "
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencodeexin[gencodeexin['genename'].str.contains('OR4F5')] #& gencodeexin['tag'].str.contains('basic')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 1 upstream, downstream, closest gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downstream 'MT-TP'\n",
      "16023\n",
      "Downstream Gene is: MIR6859-4\n"
     ]
    }
   ],
   "source": [
    "a=input()\n",
    "words = a.split()\n",
    "t=r\"^.*\\'(.*)\\'.*$\"\n",
    "tmp=re.match(t,a)\n",
    "if \"downstream\" in words:\n",
    "    pos=gencodeexin[gencodeexin['genename'].str.contains(tmp.group(1))].sort_values(by=['start']).iloc[0]['end']\n",
    "    print(pos)\n",
    "    #print(gencodeexin[gencodeexin['end']<pos].sort_values(by=['end']))\n",
    "    dgene=gencodeexin[gencodeexin['start']>pos].sort_values(by=['start']).iloc[0]['genename']\n",
    "    print(\"Downstream Gene is:\",dgene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upstream 'MT-TP'\n",
      "15956\n",
      "Upstream Gene is: MT-TT\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "a=input()\n",
    "words = a.split()\n",
    "t=r\"^.*\\'(.*)\\'.*$\"\n",
    "tmp=re.match(t,a)\n",
    "if \"upstream\" in words:\n",
    "    pos=gencodeexin[gencodeexin['genename'].str.contains(tmp.group(1))].sort_values(by=['start']).iloc[0]['start']\n",
    "    print(pos)\n",
    "    #print(gencodeexin[gencodeexin['end']<pos].sort_values(by=['end']))\n",
    "    upgene=gencodeexin[gencodeexin['end']<pos].sort_values(by=['start']).iloc[-1]['genename']\n",
    "    print(\"Upstream Gene is:\",upgene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ";;\n"
     ]
    }
   ],
   "source": [
    "a=input()\n",
    "words = a.split()\n",
    "t=r\"^.*\\'(.*)\\'.*$\"\n",
    "tmp=re.match(t,a)\n",
    "if \"closest\" in words:\n",
    "    starpos=gencodeexin[gencodeexin['genename'].str.contains(tmp.group(1))].sort_values(by=['start']).iloc[0]['start']\n",
    "    #print(gencodeexin[gencodeexin['end']<pos].sort_values(by=['end']))\n",
    "    upgenepos=gencodeexin[gencodeexin['end']<pos].sort_values(by=['start']).iloc[-1]['end']\n",
    "    upgene = gencodeexin[gencodeexin['end']<pos].sort_values(by=['start']).iloc[-1]['genename']\n",
    "    mdistance=starpos-upgenepos\n",
    "    endpos=gencodeexin[gencodeexin['genename'].str.contains(tmp.group(1))].sort_values(by=['start']).iloc[0]['end']\n",
    "    dgenepos=gencodeexin[gencodeexin['start']>pos].sort_values(by=['start']).iloc[0]['start']\n",
    "    dgene=gencodeexin[gencodeexin['start']>pos].sort_values(by=['start']).iloc[0]['genename']\n",
    "    if mdistance > (dgenepos-endpos):\n",
    "        print(\"Closest Gene is:\",upgene)\n",
    "    else:\n",
    "        print(\"Closest Gene is:\",dgene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 2, how many transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mant transcripts 'MIR6859-4'\n",
      "MIR6859-4\n",
      "1971419     None\n",
      "1971420    basic\n",
      "1971421    basic\n",
      "Name: tag, dtype: object\n"
     ]
    }
   ],
   "source": [
    "a=input()\n",
    "words = a.split()\n",
    "t=r\"^.*\\'(.*)\\'.*$\"\n",
    "tmp=re.match(t,a)\n",
    "if \"many\" and \"transcripts\" in words:\n",
    "    trs=gencodeexin[gencodeexin['genename'].str.contains(tmp.group(1))].sort_values(by=['start'])\n",
    "    c=trs.tag#.attribute#.transcript#.unique()\n",
    "    print(tmp.group(1))\n",
    "    print(c)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 3, what is in this position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m    728\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 729\u001b[0;31m                 \u001b[0mident\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreply\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdin_socket\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    730\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/jupyter_client/session.py\u001b[0m in \u001b[0;36mrecv\u001b[0;34m(self, socket, mode, content, copy)\u001b[0m\n\u001b[1;32m    802\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 803\u001b[0;31m             \u001b[0mmsg_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_multipart\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    804\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mzmq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mZMQError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/zmq/sugar/socket.py\u001b[0m in \u001b[0;36mrecv_multipart\u001b[0;34m(self, flags, copy, track)\u001b[0m\n\u001b[1;32m    394\u001b[0m         \"\"\"\n\u001b[0;32m--> 395\u001b[0;31m         \u001b[0mparts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrack\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtrack\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    396\u001b[0m         \u001b[0;31m# have first part already, only loop while more to receive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mzmq/backend/cython/socket.pyx\u001b[0m in \u001b[0;36mzmq.backend.cython.socket.Socket.recv\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mzmq/backend/cython/socket.pyx\u001b[0m in \u001b[0;36mzmq.backend.cython.socket.Socket.recv\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mzmq/backend/cython/socket.pyx\u001b[0m in \u001b[0;36mzmq.backend.cython.socket._recv_copy\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/zmq/backend/cython/checkrc.pxd\u001b[0m in \u001b[0;36mzmq.backend.cython.checkrc._check_rc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-eb8d2cf0ca62>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mre\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0ma\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mwords\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34mr\"^.*\\'(.*)\\'.*$\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36mraw_input\u001b[0;34m(self, prompt)\u001b[0m\n\u001b[1;32m    702\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_ident\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    703\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_header\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 704\u001b[0;31m             \u001b[0mpassword\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    705\u001b[0m         )\n\u001b[1;32m    706\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m    732\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    733\u001b[0m                 \u001b[0;31m# re-raise KeyboardInterrupt, to truncate traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 734\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    735\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    736\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import re\n",
    "import numpy as np\n",
    "a=input()\n",
    "words = a.split()\n",
    "t=r\"^.*\\'(.*)\\'.*$\"\n",
    "tmp=re.match(t,a)\n",
    "position=gencodeexin.drop(['attribute','tag','geneid'], axis=1)\n",
    "flag=0\n",
    "maxstart=max(position['start'])\n",
    "minend=min(position['end'])\n",
    "\n",
    "for index,row in position.iterrows():\n",
    "    if int(tmp.group(1)) in range(int(row['start']),int(row['end'])) and row['feature']=='exon':\n",
    "        flag=1\n",
    "        #print('ok')\n",
    "        nucl=int(tmp.group(1))-int(row['start'])+1\n",
    "        print(nucl)\n",
    "        transcript=row['transcript']\n",
    "        print(transcript)\n",
    "        gene=row['genename']\n",
    "        print(gene)\n",
    "        chrom=row['seqname']\n",
    "        print(chrom)\n",
    "        trs=position[position['genename'].str.contains(row['genename'])].sort_values(by=['start'])\n",
    "        trs2=trs[trs['feature'].str.contains('exon')]\n",
    "        trs3=trs2[trs2['transcript'].str.contains(transcript)]\n",
    "        trs3['nrows'] = np.arange(len(trs3))\n",
    "        r=trs3.loc[trs3['start']==row['start']]['nrows']+1\n",
    "        length=row['end']-row['start']\n",
    "        nl=[]\n",
    "        for index2,row2 in trs3.iterrows():\n",
    "            if int(tmp.group(1)) in range(int(row2['start']),int(row2['end'])):\n",
    "                vals=row2['start']\n",
    "                vale=row2['end']\n",
    "            nl.append(row2['start'])\n",
    "            nl.append(row2['end'])\n",
    "        print(nl)\n",
    "        for i in range(1,len(nl)):\n",
    "            if nl[i]==vals:\n",
    "                print(nl[i]-nl[i-1])\n",
    "            if nl[i]==vale:\n",
    "                print(nl[i+1]-nl[i])\n",
    "        break;\n",
    "        \n",
    "if flag==0:\n",
    "    print(\"This is an intergenic region\")\n",
    "\n",
    "    i=j=int(tmp.group(1))\n",
    "    while i<maxstart:\n",
    "        dowgen=gencodeexin.loc[gencodeexin.end==i].genename\n",
    "        if dowgen.empty:\n",
    "            i=i+1\n",
    "        else:\n",
    "            break;\n",
    "    print(\"downstream gene is\",tuple(dowgen)[0])\n",
    "    while j>minend:\n",
    "        upgen=gencodeexin.loc[gencodeexin.start==j].genename\n",
    "        if upgen.empty:\n",
    "            j=j-1\n",
    "        else:\n",
    "            break;\n",
    "    print(\"upstream gene is\",tuple(upgen)[0])       \n",
    "        \n",
    "#print(trs3)\n",
    "#print(r)\n",
    "\n",
    "\n",
    "import requests, sys\n",
    "     \n",
    "server = \"https://rest.ensembl.org\"\n",
    "match_chr=re.match(r'chr(.*)',chrom)\n",
    "expand_head=int(tmp.group(1))+50\n",
    "expand_tail=int(tmp.group(1))-50\n",
    "\n",
    "\n",
    "ext = '/sequence/region/human/' + match_chr.group(1) + ':' + str(expand_tail) + '..'  + str(expand_head) + '?'     \n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"text/plain\"})\n",
    "     \n",
    "if not r.ok:\n",
    "    r.raise_for_status()\n",
    "    sys.exit()\n",
    "     \n",
    "#print(len(r.text))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The name of TPMT is : thiopurine S-methyltransferase\n",
      "\n",
      "\n",
      "Other names of TPMT are : ['S-adenosyl-L-methionine:thiopurine S-methyltransferase', 'thiopurine S-methyltransferase']\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "#a=input()\n",
    "a='tpmt'\n",
    "parameters = {\n",
    "  'fields': 'name,other_names',\n",
    "  'species': 'human',\n",
    "  'q' : 'symbol:' + a,\n",
    "}\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "\n",
    "#response.ok\n",
    "\n",
    "data = response.json()\n",
    "\n",
    "print (f'The name of {a.upper()} is :', (data['hits'][0]['name']))\n",
    "print('\\n')\n",
    "print (f'Other names of {a.upper()} are :',\n",
    "    data['hits'][0]['other_names']) #### κάπως να μην φαίνεται η λίστα \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Where is the location of GENE\n",
      "['Where', 'is', 'the', 'location', 'of', 'GENE']\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "a=input()\n",
    "words = a.split()\n",
    "t=r\"^.*\\'(.*)\\'.*$\"\n",
    "tmp=re.match(t,a)\n",
    "#print(tmp)\n",
    "print(words)\n",
    "#print(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome tpmt\n",
      "['chromosome', 'tpmt']\n",
      "tpmt\n",
      "The chomosome of gene <TPMT> is: chr 6\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "a=input()\n",
    "words = a.split()\n",
    "print(words)\n",
    "print(words[-1])\n",
    "a=words[-1]\n",
    "\n",
    "if 'location' in words:\n",
    "\n",
    "    if 'rs' in a:\n",
    "\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': + a ,\n",
    "            'fields': 'dbsnp',\n",
    "        }\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "        chrom = data['hits'][0]['dbsnp']['chrom']\n",
    "        start = data['hits'][0]['dbsnp']['hg19']['start']\n",
    "        end = data['hits'][0]['dbsnp']['hg19']['end']\n",
    "\n",
    "        print (f'Mutation: {a}'+'\\n'+\n",
    "               f'Chomosome: {chrom}'+'\\n'+\n",
    "               f'Start: {start}'+'\\n'\n",
    "               f'End: {end}')\n",
    "        #print(data)\n",
    "\n",
    "    elif 'ENST' in a:\n",
    "\n",
    "        parameters = {\n",
    "      'fields': 'genomic_pos,map_location',\n",
    "      'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + a}\n",
    "\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        chom=data['hits'][0]['genomic_pos']['chr']\n",
    "        start=data['hits'][0]['genomic_pos']['start']\n",
    "        end=data['hits'][0]['genomic_pos']['end']\n",
    "        map_loc=data['hits'][0]['map_location']\n",
    "\n",
    "        print(f'The chomosome of transcript <{a}> is:', 'chr' + chom +'\\n'+\n",
    "              f'Start: {start}'+'\\n'+\n",
    "              f'End : {end}'+'\\n'+ \n",
    "              f'Map location: {map_loc}')\n",
    "\n",
    "    else:\n",
    "\n",
    "        parameters = {\n",
    "      'fields': 'genomic_pos,map_location',\n",
    "      'species': 'human',\n",
    "      'q' : 'symbol:' + a }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "        chrom=data['hits'][0]['genomic_pos']['chr']\n",
    "        start=data['hits'][0]['genomic_pos']['start']\n",
    "        end=data['hits'][0]['genomic_pos']['end']\n",
    "        map_loc=data['hits'][0]['map_location']\n",
    "\n",
    "        print(f'The chomosome of gene <{a.upper()}> is:', 'chr' + 'chr' + chrom +'\\n'+\n",
    "              f'Start: {start}' +'\\n'+\n",
    "              f'End : {end}' +'\\n'+\n",
    "              f'Map location: {map_loc}')\n",
    "        \n",
    "elif 'chromosome' in words:\n",
    "    \n",
    "    if 'rs' in a:\n",
    "\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': + a,\n",
    "            'fields': 'dbsnp',\n",
    "        }\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "        chrom = data['hits'][0]['dbsnp']['chrom']\n",
    "\n",
    "        print (f'Mutation: {a}'+'\\n'+\n",
    "               f'Chomosome: {chrom}')\n",
    "        #print(data)\n",
    "\n",
    "    elif 'ENST' in a:\n",
    "\n",
    "        parameters = {\n",
    "      'fields': 'genomic_pos',\n",
    "      'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + a}\n",
    "\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        chom=data['hits'][0]['genomic_pos']['chr']\n",
    "\n",
    "        print(f'The chomosome of transcript <{a}> is:', chom)\n",
    "    \n",
    "    else:\n",
    "\n",
    "        parameters = {\n",
    "      'fields': 'genomic_pos',\n",
    "      'species': 'human',\n",
    "      'q' : 'symbol:' + a \n",
    "        }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "        chrom=data['hits'][0]['genomic_pos']['chr']\n",
    "\n",
    "        print(f'The chomosome of gene <{a.upper()}> is:' + ' ' + 'chr',chrom)\n",
    "\n",
    "elif 'size' in words:\n",
    "    \n",
    "    if 'rs' in a:\n",
    "    \n",
    "            url = 'http://myvariant.info/v1/query'\n",
    "            parameters = {\n",
    "                'q': + a,\n",
    "                'fields': 'dbsnp',\n",
    "            }\n",
    "\n",
    "            response = requests.get(url, params=parameters)\n",
    "            data = response.json()\n",
    "            chrom = data['hits'][0]['dbsnp']['chrom']\n",
    "            start = data['hits'][0]['dbsnp']['hg19']['start']\n",
    "            end = data['hits'][0]['dbsnp']['hg19']['end']\n",
    "\n",
    "            if start and end:\n",
    "                print (f'Mutation: {a}'+'\\n'+ \n",
    "                   f'Size:', end-start)\n",
    "            elif start or end :\n",
    "                print (f'Mutation: {a}'+'\\n'+\n",
    "                   f'Point of mutation is:', end or start)\n",
    "            else:\n",
    "                raise Exception(f'Your mutation {a} does not exist in dbsnp')\n",
    "\n",
    "    elif 'ENST' in a:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'genomic_pos',\n",
    "        'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + a\n",
    "        }\n",
    "\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        start=data['hits'][0]['genomic_pos']['start']\n",
    "        end=data['hits'][0]['genomic_pos']['end']\n",
    "\n",
    "        print(f'The size of transcript <{a}> is:', 'Size = ', end-start)\n",
    "\n",
    "\n",
    "    else:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'genomic_pos',\n",
    "        'species': 'human',\n",
    "        'q' : 'symbol:' + a \n",
    "        }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "        start=data['hits'][0]['genomic_pos']['start']\n",
    "        end=data['hits'][0]['genomic_pos']['end']\n",
    "\n",
    "    print(f'The size of gene <{a.upper()}> is:', 'Length =', end-start)\n",
    "\n",
    "\n",
    "elif 'start' in words:\n",
    "    \n",
    "    if 'rs' in a:\n",
    "                       \n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': + a ,\n",
    "            'fields': 'dbsnp',\n",
    "        }\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "        start = data['hits'][0]['dbsnp']['hg19']['start']\n",
    "\n",
    "        if start:\n",
    "            print (f'The start of Mutation: {a} is' ,start)\n",
    "\n",
    "        else:\n",
    "            raise Exception(f'Your mutation {a} does not have start point in dbsnp')\n",
    "\n",
    "    elif 'ENST' in a:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'genomic_pos',\n",
    "        'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + a\n",
    "        }\n",
    "\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        start=data['hits'][0]['genomic_pos']['start']\n",
    "\n",
    "        print(f'The start of transcript <{a}> is:', 'Start = ', start)\n",
    "\n",
    "\n",
    "    else:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'genomic_pos',\n",
    "        'species': 'human',\n",
    "        'q' : 'symbol:' + a \n",
    "         }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        #response.ok\n",
    "\n",
    "        data = response.json()\n",
    "        start=data['hits'][0]['genomic_pos']['start']                   \n",
    "\n",
    "        print(f'The start point of gene <{a.upper()}> is:', 'Start =', start)\n",
    "    \n",
    "elif 'end' in words:\n",
    "        \n",
    "    if 'rs' in a:\n",
    "    \n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': + a,\n",
    "            'fields': 'dbsnp',\n",
    "        }\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "\n",
    "        end = data['hits'][0]['dbsnp']['hg19']['end']\n",
    "\n",
    "        if end:\n",
    "            print (f'The end Mutation: {a} is ', end)\n",
    "\n",
    "        else:\n",
    "            raise Exception(f'Your mutation {a} does not have end point in dbsnp')\n",
    "\n",
    "    elif 'ENST' in a:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'genomic_pos',\n",
    "        'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + a\n",
    "        }\n",
    "\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        end=data['hits'][0]['genomic_pos']['end']\n",
    "\n",
    "        print(f'The end of transcript <{a}> is:', 'End = ', end)\n",
    "\n",
    "\n",
    "    else:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'genomic_pos',\n",
    "        'species': 'human',\n",
    "        'q' : 'symbol:' + a \n",
    "                }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "\n",
    "        data = response.json()\n",
    "        end=data['hits'][0]['genomic_pos']['end']\n",
    "\n",
    "        print(f'The end of gene <{a.upper()}> is:', 'End  =', end)\n",
    "\n",
    "else:\n",
    "    raise Exception('Please try again and specify your question with words such as: <location, chromosome, size, start or end>')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 6\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "refseq TPMT\n",
      "['refseq', 'TPMT']\n",
      "TPMT\n",
      "TPMT\n",
      "NM_000367.4\n",
      "       Gene  Score         Rnas         CDS    Principal\n",
      "22347  TPMT   7172  NM_000367.4  CCDS4543.1  PRINCIPAL:1\n",
      "PRINCIPAL:1\n",
      "The transcript: NM_000367.4 is primary\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "appris = pd.read_table(\"/home/thodoris/Desktop/appris_data.principal.txt\",names = ['Gene', 'Score', 'Rnas', 'CDS', 'Principal'])\n",
    "a=input()\n",
    "words = a.split()\n",
    "print(words)\n",
    "print(words[-1])\n",
    "a=words[-1]\n",
    "###########---two ways---#########\n",
    "\n",
    "#aaa=re.sub(r'\\?', '',a )\n",
    "\n",
    "a=a.replace('?', '')\n",
    "print(a)\n",
    "\n",
    "\n",
    "if 'refseq' in words:\n",
    "    parameters = {\n",
    "        'fields': 'refseq',\n",
    "        'species': 'human',\n",
    "        'q' : 'symbol:' + a \n",
    "                }\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "\n",
    "    data = response.json()\n",
    "    transcripts=data['hits'][0]['refseq']['rna']\n",
    "    gtable_not_CDS=appris[appris['CDS'].str.contains('-')==False]\n",
    "    gene_df=gtable_not_CDS[gtable_not_CDS['Gene'].str.contains(a.upper())]\n",
    "    #print(gene_df)\n",
    "    for trc in gene_df['Rnas']:\n",
    "        if trc in transcripts:\n",
    "            print(trc)\n",
    "            #df.loc[(df['age'] == 21) & df['favorite_color'].isin(array)]\n",
    "            row_df=gene_df.loc[(gene_df['Rnas'] == trc)] #& gene_df['Principal'] == 'PRINCIPAL:1')]\n",
    "            print(row_df)\n",
    "            for prc in row_df['Principal']:\n",
    "                print(prc)\n",
    "                if prc == 'PRINCIPAL:1':\n",
    "                    print(f'The transcript: {trc} is primary' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many exons does <TPMT> have ?\n",
      "TPMT\n",
      " Gene TPMT in chromosome:6, contains 9 exons.\n"
     ]
    }
   ],
   "source": [
    "####------FINAL-----7\n",
    "\n",
    "import re\n",
    "a=input()\n",
    "m = re.search(r\"\\<([A-Za-z0-9_]+)\\>\", a)\n",
    "query=m.group(1)\n",
    "words=a.split()\n",
    "print(query)\n",
    "\n",
    "if query is not None:\n",
    "    \n",
    "    if 'ENST' in query:\n",
    "\n",
    "        parameters = {\n",
    "        'fields': 'exons',\n",
    "        'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + query \n",
    "                }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "        #print(parameters)\n",
    "        response.ok\n",
    "\n",
    "        chrm=data['hits'][0]['exons'][0]['chr']\n",
    "        pos=data['hits'][0]['exons'][0]['position']\n",
    "        print(f'Chromosome:{chrm} contains {len(pos)}' + '\\n',\n",
    "              'The Start/End positions of Transcipts {query}' + '\\n', *pos, sep='\\n')\n",
    "        \n",
    "    elif 'many' in words: \n",
    "\n",
    "        parameters = {\n",
    "                'fields': 'exons',\n",
    "                'species': 'human',\n",
    "                'q' : 'symbol:' + query \n",
    "                    }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "        #print(parameters)\n",
    "        response.ok\n",
    "        #print(data['hits'][0]['exons'])\n",
    "        chrm=data['hits'][0]['exons'][0]['chr']\n",
    "        pos=data['hits'][0]['exons'][0]['position']\n",
    "        print(f' Gene {query} in chromosome:{chrm}, contains {len(pos)} exons.')\n",
    "\n",
    "    elif 'locations' in words:\n",
    "        parameters = {\n",
    "                'fields': 'exons',\n",
    "                'species': 'human',\n",
    "                'q' : 'symbol:' + query \n",
    "                    }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "        #response.ok\n",
    "        chrm=data['hits'][0]['exons'][0]['chr']\n",
    "        pos=data['hits'][0]['exons'][0]['position']\n",
    "        lens=[b-i for i,b in pos]\n",
    "        mega_list=list(zip(pos,lens))\n",
    "        mega_list.sort(key=lambda x: x[1], reverse=True)\n",
    "        #mega_list.sort(key=operator.itemgetter(1))\n",
    "        print(f'Chromosome:{chrm}, contains {len(pos)} Exons'+ '\\n',\n",
    "              'The Start/End positions and Length:' + '\\n', *mega_list, sep='\\n',)\n",
    "    else:\n",
    "        raise Exception('Please define your query with official gene symbol or Ensembl transcript id')\n",
    "\n",
    "else:\n",
    "    raise Exception('Your input in none, define your query with official gene symbol or Ensembl transcript id ')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " # Question 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TPMT\n",
      "The transcript: NM_000367.4 is primary\n",
      "NM_000367\n",
      "[(0, 'cdsend', 18149127), (0, 'cdsstart', 18130667), (0, 'transcript', 'NM_000367'), (1, 'cdsend', 18149127), (1, 'cdsstart', 18130667), (1, 'transcript', 'NM_001346817'), (2, 'cdsend', 18149127), (2, 'cdsstart', 18130667), (2, 'transcript', 'NM_001346818')]\n",
      "NM_000367\n",
      "(0, 'transcript', 'NM_000367')\n",
      "ok\n",
      "The position of this <NM_000367> fucking primay transcript is:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import re\n",
    "import pandas as pd\n",
    "a='Where is the location of the CDS of the <TPMT> ?'\n",
    "#a=input()\n",
    "m = re.search(r\"\\<([A-Za-z0-9_]+)\\>\", a)\n",
    "query=m.group(1)\n",
    "words=a.split()\n",
    "appris = pd.read_table(\"/home/thodoris/Desktop/appris_data.principal.txt\",names = ['Gene', 'Score', 'Rnas', 'CDS', 'Principal'])\n",
    "\n",
    "m = re.search(r\"\\<([A-Za-z0-9_]+)\\>\", a)\n",
    "query=m.group(1)\n",
    "words=a.split()\n",
    "print(query)\n",
    " \n",
    "if query is not None:\n",
    "    \n",
    "    if 'CDS' in words:\n",
    "        parameters = {\n",
    "                'fields': 'refseq,exons',\n",
    "                'species': 'human',\n",
    "                'q' : 'symbol:' + query \n",
    "                    }\n",
    "        url = 'http://mygene.info/v3/query'\n",
    "\n",
    "        response = requests.get(url, params=parameters)\n",
    "        data = response.json()\n",
    "\n",
    "        tra_1=data['hits'][0]['refseq']['rna']\n",
    "        tra_2=data['hits'][0]['exons']\n",
    "        #print(data)\n",
    "        array=['transcript','cdsend','cdsstart']\n",
    "\n",
    "        for nms in tra_1:\n",
    "            row=appris[appris['Rnas']== nms]\n",
    "            break\n",
    "        final_df=row[row['CDS'].str.contains('-')==False]\n",
    "            \n",
    "        for principal in final_df['Principal']:\n",
    "            if principal == 'PRINCIPAL:1':\n",
    "                print(f'The transcript: {nms} is primary' )\n",
    "                break\n",
    "\n",
    "            else:\n",
    "                print('There are no primaires')\n",
    "\n",
    "        NM = (re.search(r\"([A-Za-z0-9_]+)\", nms).group(1))\n",
    "        print(NM)\n",
    "        #transcripts=[list((k,v)) for i in tra_2 for k,v in i.items() if k in array]\n",
    "        transcripts=[(i,k,v) for i,j in enumerate(tra_2) for k,v in j.items() if k in array]\n",
    "        print(transcripts)\n",
    "        for elem in transcripts:\n",
    "            if NM == elem[2]:\n",
    "                print(NM)\n",
    "                print(elem)\n",
    "                break\n",
    "       # print(elem)\n",
    "        print('ok')\n",
    "        print(f'The position of this <{NM}> primay transcript is:' + '\\n', )\n",
    "         #    )\n",
    "\n",
    "\n",
    "    else:\n",
    "        raise Exception('Please define your query with official gene symbol or Ensembl transcript id')\n",
    "\n",
    "else:\n",
    "    raise Exception('Your input in none, define your query with official gene symbol or Ensembl transcript id ')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 9\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['What', 'is', 'the', 'primary', 'transcript', 'of', 'the', 'gene', '<TPMT>', '?']\n",
      "TPMT\n",
      "['NM_000367.4', 'NM_001346817.1', 'NM_001346818.1']\n",
      "NM_000367.4\n",
      "       Gene  Score         Rnas         CDS    Principal\n",
      "22347  TPMT   7172  NM_000367.4  CCDS4543.1  PRINCIPAL:1\n",
      "PRINCIPAL:1\n",
      "The transcripts: NM_000367.4,NM_001346817.1,NM_001346818.1, the primary one, is NM_000367.4\n",
      " Refseq_ID: NM_000367.4\n",
      " Ensembl_ID: ENST00000309983\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "import pandas as pd\n",
    "appris = pd.read_table(\"/home/thodoris/Desktop/appris_data.principal.txt\",names = ['Gene', 'Score', 'Rnas', 'CDS', 'Principal'])\n",
    "a='What is the primary transcript of the gene <TPMT> ?'\n",
    "#a=input()\n",
    "words = a.split()\n",
    "print(words)\n",
    "gene = (re.search(r\"\\<([A-Za-z0-9_]+)\\>\", a).group(1))\n",
    "print(gene)\n",
    "\n",
    "parameters = {\n",
    "    'fields': 'refseq,ensembl.transcript',\n",
    "    'species': 'human',\n",
    "    'q' : 'symbol:' + gene \n",
    "            }\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "\n",
    "data = response.json()\n",
    "\n",
    "transcripts=data['hits'][0]['refseq']['rna']\n",
    "print(transcripts)\n",
    "\n",
    "gtable_not_CDS=appris[appris['CDS'].str.contains('-')==False]\n",
    "gene_df=gtable_not_CDS[gtable_not_CDS['Gene'].str.contains(gene.upper())]\n",
    "#print(gene_df)\n",
    "for trc in gene_df['Rnas']:\n",
    "    if trc in transcripts:\n",
    "        print(trc)\n",
    "        #df.loc[(df['age'] == 21) & df['favorite_color'].isin(array)]\n",
    "        row_df=gene_df.loc[(gene_df['Rnas'] == trc)] #& gene_df['Principal'] == 'PRINCIPAL:1')]\n",
    "        print(row_df)\n",
    "        for prc in row_df['Principal']:\n",
    "            print(prc)\n",
    "            if prc == 'PRINCIPAL:1':\n",
    "                ensembl_id=data['hits'][0]['ensembl']['transcript']\n",
    "                print(f\"The transcripts: {','.join(transcripts)}, the primary one, is {trc}\"+'\\n', f'Refseq_ID: {trc}' + '\\n', f'Ensembl_ID: {ensembl_id}' )\n",
    "#print(data)\n",
    "#print(data['hits'][0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "import pandas as pd\n",
    "appris = pd.read_table(\"/home/thodoris/Desktop/appris_data.principal.txt\",names = ['Gene', 'Score', 'Rnas', 'CDS', 'Principal'])\n",
    "a='What is the primary transcript of the gene <OR4F5> ?'\n",
    "#a=input()\n",
    "words = a.split()\n",
    "print(words)\n",
    "gene = (re.search(r\"\\<([A-Za-z0-9_]+)\\>\", a).group(1))\n",
    "print(gene)\n",
    "\n",
    "parameters = {\n",
    "    'fields': 'refseq,ensembl.transcript',\n",
    "    'species': 'human',\n",
    "    'q' : 'symbol:' + gene \n",
    "            }\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "\n",
    "data = response.json()\n",
    "\n",
    "transcripts=data['hits'][0]['refseq']['rna']\n",
    "\n",
    "gtable_not_CDS=appris[appris['CDS'].str.contains('-')==False]\n",
    "gene_df=gtable_not_CDS[gtable_not_CDS['Gene'].str.contains(gene.upper())]\n",
    "#print(gene_df)\n",
    "for trc in gene_df['Rnas']:\n",
    "    if trc in transcripts:\n",
    "        print(trc)\n",
    "        #df.loc[(df['age'] == 21) & df['favorite_color'].isin(array)]\n",
    "        row_df=gene_df.loc[(gene_df['Rnas'] == trc)] #& gene_df['Principal'] == 'PRINCIPAL:1')]\n",
    "        print(row_df)\n",
    "        for prc in row_df['Principal']:\n",
    "            print(prc)\n",
    "            if prc == 'PRINCIPAL:1':\n",
    "                ensembl_id=data['hits'][0]['ensembl']['transcript']\n",
    "                print(f\"The transcripts: {transcripts}, the primary one, is {trc}\"+'\\n', f'Refseq_ID: {trc}' + '\\n', f'Ensembl_ID: {ensembl_id}' )\n",
    "                break\n",
    "            else:\n",
    "                print('There are no primary transcritps')\n",
    "\n",
    "############\n",
    "#σου γυρνάει 2 ENSTs  γτ το κάθε ένα ανήκει σε άλλο γονίδιο πχ : OR4F5-201 , : OR4F5-202\n",
    "#αλλά οι μαλάκες δεν το έχουν κάνει διαχρισμό στο mygene.info χειροκίνητα το βρήκα \n",
    "############\n",
    "ensembl_id=data['hits'][0]['ensembl']['transcript']\n",
    "\n",
    "for enst in ensembl_id:\n",
    "    parameters = {\n",
    "        'fields': 'symbol,ensembl.transcript',\n",
    "        'species': 'human',\n",
    "        'q' : 'ensembl.transcript:' + enst \n",
    "                }\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "\n",
    "    data = response.json()\n",
    "    #print(data['hits'][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['In', 'which', 'gene', 'does', '<ENST00000444521.6>', 'belong?']\n",
      "ENST00000444521.6\n",
      "ENST00000444521\n",
      "{'max_score': 19.60245, 'took': 31, 'total': 1, 'hits': [{'_id': '127281', '_score': 19.60245, 'genomic_pos': [{'chr': '1', 'end': 2591469, 'ensemblgene': 'ENSG00000157870', 'start': 2586491, 'strand': 1}, {'chr': 'CHR_HSCHR1_1_CTG3', 'end': 2591469, 'ensemblgene': 'ENSG00000275125', 'start': 2586491, 'strand': 1}], 'name': 'peroxiredoxin like 2B', 'symbol': 'PRXL2B'}]}\n",
      "The gene-name of transcript <ENST00000444521> is <peroxiredoxin like 2B> with gene-symbol <PRXL2B>.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "a='In which gene does <ENST00000444521.6> belong?'\n",
    "#a=input()\n",
    "############ trp means (transcript, rs, position)\n",
    "#################################################\n",
    "trp = (re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "words = a.split()\n",
    "print(words)\n",
    "print(trp)\n",
    "\n",
    "\n",
    "if 'rs' in a:\n",
    "\n",
    "    url = 'http://myvariant.info/v1/query'\n",
    "    parameters = {\n",
    "        'q': trp,\n",
    "        'fields': 'snpeff', ####### για να βρω την κατάλληλη βάση να ψάξω πρώτα έβαλα 'all'########\n",
    "    }\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data = response.json()\n",
    "    gene=data['hits'][0]['snpeff']['ann']['genename']\n",
    "\n",
    "\n",
    "    print (f'The mutation: {trp} is located on Gene {gene}.')\n",
    "\n",
    "elif 'ENST' in a:\n",
    "    \n",
    "    transcript= ((re.search(r\"\\<([A-Za-z0-9_]+)\", a)).group(1))\n",
    "    print(transcript)\n",
    "    parameters = {\n",
    "        \n",
    "    'fields': 'genomic_pos,symbol,name',\n",
    "    'species': 'human',\n",
    "    'q' : 'ensembl.transcript:' + transcript\n",
    "    }\n",
    "\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data = response.json()\n",
    "    print(data)\n",
    "    name=data['hits'][0]['name']\n",
    "    symbol=data['hits'][0]['symbol']\n",
    "\n",
    "\n",
    "    print(f'The gene-name of transcript <{transcript}> is <{name}> with gene-symbol <{symbol}>.')\n",
    "\n",
    "\n",
    "else:\n",
    "    print('Specify your question')\n",
    "\n",
    "   \n",
    "    \n",
    "    \n",
    "######### από την ερώτηση αυτή έμεινε το 3 δλδ αν βάλουμε θέση, θα βρεί γονίδιο?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['In', 'which', 'pathways', 'is', 'the', 'gene', '<TP73>', 'involved', '?']\n",
      "TP73\n",
      "According to biocarta: ['atm signaling pathway']\n",
      "According to kegg: ['p53 signaling pathway - Homo sapiens (human)', 'Hippo signaling pathway - Homo sapiens (human)', 'Neurotrophin signaling pathway - Homo sapiens (human)', 'Measles - Homo sapiens (human)']\n",
      "According to netpath: ['Alpha6Beta4Integrin', 'TGF_beta_Receptor']\n",
      "According to pid: ['E2F transcription factor network', 'Direct p53 effectors', 'p73 transcription factor network']\n",
      "According to reactome: ['Apoptosis', 'Intrinsic Pathway for Apoptosis', 'Activation of BH3-only proteins', 'Activation of PUMA and translocation to mitochondria', 'Generic Transcription Pathway', 'Transcriptional Regulation by TP53', 'Programmed Cell Death', 'Regulation of TP53 Activity', 'TP53 Regulates Transcription of Cell Death Genes', 'TP53 Regulates Transcription of Genes Involved in Cytochrome C Release', 'TP53 regulates transcription of several additional cell death genes whose specific roles in p53-dependent apoptosis remain uncertain', 'TP53 Regulates Transcription of Caspase Activators and Caspases', 'TP53 Regulates Transcription of Death Receptors and Ligands', 'Regulation of TP53 Activity through Association with Co-factors', 'RNA Polymerase II Transcription', 'Gene expression (Transcription)', 'Transcriptional regulation by RUNX1', 'RUNX1 regulates transcription of genes involved in differentiation of HSCs']\n",
      "According to wikipathways: ['TP53 Network', 'ATM Signaling Pathway', 'Apoptosis', 'Sudden Infant Death Syndrome (SIDS) Susceptibility Pathways', 'DNA Damage Response (only ATM dependent)']\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import re\n",
    "from collections import defaultdict\n",
    "a=\"In which pathways is the gene <TP73> involved ?\"\n",
    "\n",
    "gene = (re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "words = a.split()\n",
    "print(words)\n",
    "print(gene)\n",
    "\n",
    "parameters = {\n",
    "  'fields': 'pathway',\n",
    "  'species': 'human',\n",
    "  'q' : 'symbol:' + gene,\n",
    "}\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "data = response.json()\n",
    "dicts=data['hits'][0]['pathway']\n",
    "subdicts=[v for k,v in dicts.items()]\n",
    "\n",
    "dict_list=data['hits'][0]['pathway']\n",
    "\n",
    " \n",
    "\n",
    "dbs_names = []\n",
    "for k,v in dict_list.items():\n",
    "    if isinstance(v,dict):\n",
    "        dbs_names.append((k,v['name']))\n",
    "       # print(f'According to {k}:',v['name'])\n",
    "    elif isinstance(v,list):\n",
    "        for sub_dict in v:\n",
    "            dbs_names.append((k,sub_dict['name']))\n",
    "\n",
    "paths_by_db = defaultdict(list)\n",
    "for db, path in dbs_names:\n",
    "    paths_by_db[db].append(path)\n",
    "    \n",
    "for db, path in paths_by_db.items():\n",
    "    print (f'According to {db}: {path}')\n",
    "#print(dbs2_names)\n",
    "\n",
    "\n",
    "#print(f'According:' + '\\n', *dbs2_names, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Which', 'proteins', 'are', 'transcribed', 'from', 'the', '<TPMT>', '?']\n",
      "TPMT\n",
      "ok\n",
      "[['NM_000367.4 ---> NP_000358.1', 'Primary'], ['NM_001346818.1 ---> NP_001333747.1', 'Not'], ['NM_001346817.1 ---> NP_001333746.1', 'Not']]\n",
      "RNAS / Proteins:\n",
      "\n",
      "['NM_000367.4 ---> NP_000358.1', 'Primary']\n",
      "['NM_001346818.1 ---> NP_001333747.1', 'Not']\n",
      "['NM_001346817.1 ---> NP_001333746.1', 'Not']\n"
     ]
    }
   ],
   "source": [
    "a=\"Which proteins are transcribed from the <TPMT> ?\"\n",
    "import re\n",
    "import requests\n",
    "from operator import itemgetter\n",
    "trp = (re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "words = a.split()\n",
    "print(words)\n",
    "print(trp)\n",
    "if 'ENST' in trp:\n",
    "    print('ok')\n",
    "    transcript= (re.search(r\"\\<([A-Za-z0-9_]+)\", a).group(1))\n",
    "    print(transcript)\n",
    "    parameters = {\n",
    "    'fields': 'refseq',\n",
    "    'species': 'human',\n",
    "    'q' : 'ensembl.transcript:' + transcript\n",
    "    }\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data = response.json()\n",
    "    esnts=data['hits'][0]['refseq']['translation']\n",
    "    rnas_list_1=data['hits'][0]['refseq']['translation']\n",
    "    print(rnas_list_1)\n",
    "\n",
    "    if len(rnas_list_1) > 2:\n",
    "        b=[[sub_dict['rna']+' '+'--->'+' '+sub_dict['protein'],is_primary(sub_dict['rna'])] for sub_dict in rnas_list_1]\n",
    "        print(b)\n",
    "\n",
    "        print(f'RNAS / Proteins:' + '\\n', *b, sep='\\n')\n",
    "    else:\n",
    "        b2=[rnas_list_1['rna'],rnas_list_1['protein'] + ' ' + '--->' + ' ' + is_primary(rnas_list_1['rna'])]\n",
    "        #print(b2)\n",
    "        print(f'RNAS / Proteins:' + '\\n', f' {b2}')\n",
    "\n",
    "\n",
    "elif 'NM' in trp:\n",
    "\n",
    "    transcript_2= ((re.search(r\"\\<([A-Za-z0-9_]+)\", a)).group(1))\n",
    "\n",
    "    parameters = {\n",
    "      'fields': 'refseq',\n",
    "      'species': 'human',\n",
    "      'q' : 'refseq:' + transcript_2,\n",
    "    }\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data=response.json()\n",
    "    translation_2=data['hits'][0]['refseq']['translation']\n",
    "    #listA2=[(sub_dict['rna']+' '+'--->'+' '+sub_dict['protein']) for sub_dict in translation_2]\n",
    "    print(translation_2)\n",
    "    #print(f'RNAS / Proteins:' + '\\n', *listA2, sep='\\n')\n",
    "\n",
    "    if len(translation_2) > 2:\n",
    "        rps_1=[[sub_dict['rna']+' '+'--->'+' '+sub_dict['protein'],is_primary(sub_dict['rna'])] for sub_dict in translation_2]\n",
    "        print(rps_1)\n",
    "        print(f'RNAS / Proteins:' + '\\n', *rps_1, sep='\\n')\n",
    "    else:\n",
    "        rp_s=[translation_2['rna'],translation_2['protein'] + ' ' + '--->' + ' ' + is_primary(translation_2['rna'])]\n",
    "        #print(b2)\n",
    "        print(f'RNAS / Proteins:' + '\\n', f' {rp_s}')\n",
    "\n",
    "\n",
    "\n",
    "else:\n",
    "    print('ok')\n",
    "    gene_id=(re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "    parameters = {\n",
    "      'fields': 'refseq',\n",
    "      'species': 'human',\n",
    "      'q' : 'symbol:' + gene_id,\n",
    "    }\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "\n",
    "    data = (requests.get(url, params=parameters).json())\n",
    "    rnas_proteins=data['hits'][0]['refseq']['translation']\n",
    "    rnas_list=data['hits'][0]['refseq']['rna']\n",
    "\n",
    "    if len(rnas_proteins) > 2:\n",
    "        rps1=[[sub_dict['rna']+' '+'--->'+' '+sub_dict['protein'],is_primary(sub_dict['rna'])] for sub_dict in rnas_proteins]\n",
    "        print(rps1)\n",
    "        print(f'RNAS / Proteins:' + '\\n', *rps1, sep='\\n')\n",
    "\n",
    "    else:\n",
    "        rps=[rnas_proteins['rna'],rnas_proteins['protein'] + ' ' + '--->' + ' ' + is_primary(rnas_proteins['rna'])]\n",
    "        #print(b2)\n",
    "        print(f'RNAS / Proteins:' + '\\n', f' {rps}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['What', 'is', 'the', 'function', 'of', 'the', 'gene', '<TPMT>', '?']\n",
      "TPMT\n",
      " The function of the gene TPMT:\n",
      " This gene encodes the enzyme that metabolizes thiopurine drugs via S-adenosyl-L-methionine as the S-methyl donor and S-adenosyl-L-homocysteine as a byproduct. Thiopurine drugs such as 6-mercaptopurine are used as chemotherapeutic agents. Genetic polymorphisms that affect this enzymatic activity are correlated with variations in sensitivity and toxicity to such drugs within individuals, causing thiopurine S-methyltransferase deficiency. Related pseudogenes have been identified on chromosomes 3, 18 and X. [provided by RefSeq, Aug 2014].\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import re\n",
    "\n",
    "a=\"What is the function of the gene <TPMT> ?\"\n",
    "\n",
    "gene = (re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "words = a.split()\n",
    "print(words)\n",
    "print(gene)\n",
    "\n",
    "parameters = {\n",
    "  'fields': 'summary',\n",
    "  'species': 'human',\n",
    "  'q' : 'symbol:' + gene,\n",
    "}\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "data = response.json()\n",
    "\n",
    "function=data['hits'][0]['summary']\n",
    "\n",
    "print(f' The function of the gene {gene}:' + '\\n',function)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Organisms:\n",
      "\n",
      "Homo sapiens\n",
      "Mus musculus\n",
      "Rattus norvegicus\n"
     ]
    }
   ],
   "source": [
    "a='Which organisms have a homologous gene to <TPMT> ?'\n",
    "taxonomy = pd.read_csv(\"/home/thodoris/Desktop/taxids.csv\", sep=',', header=1)\n",
    "gene = (re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "\n",
    "parameters = {\n",
    "  'fields': '',\n",
    "  'q' : 'symbol:' + gene\n",
    "}\n",
    "\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "data = response.json()\n",
    "mulsti_dict_list=data['hits']\n",
    "taxids_list=[sub_dict['taxid'] for sub_dict in mulsti_dict_list]\n",
    "\n",
    "#######---First---Way---###########\n",
    "#organisms=[taxonomy.loc[taxonomy['tax_id'] == ids, 'Taxonomy'].iloc[0] for tax_id in taxonomy['tax_id'] for ids in taxids_list if tax_id == ids]\n",
    "\n",
    "##########-----Second---Way---#####\n",
    "organisms=[taxonomy.query('tax_id==' + str(ids))['Taxonomy'].iloc[0] for tax_id in taxonomy['tax_id'] for ids in taxids_list if tax_id == ids]\n",
    "\n",
    "\n",
    "print(f'Organisms:' + '\\n', *organisms, sep='\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Questio 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The type of the gene <TPMT> is: protein-coding\n"
     ]
    }
   ],
   "source": [
    "a=\"What type is the gene <TPMT> ?\"\n",
    "\n",
    "gene = (re.search(r\"\\<([A-Za-z0-9_.]+)\\>\", a).group(1))\n",
    "\n",
    "parameters = {\n",
    "  'fields': 'type_of_gene',\n",
    "  'species' : 'human',\n",
    "  'q' : 'symbol:' + gene\n",
    "}\n",
    "\n",
    "url = 'http://mygene.info/v3/query'\n",
    "\n",
    "response = requests.get(url, params=parameters)\n",
    "data = response.json()\n",
    "type_of_gene=data['hits'][0]['type_of_gene']\n",
    "print(f'The type of the gene <{gene}> is: {type_of_gene}')\n",
    "\n",
    "\n",
    "########## εδώ πρέπει να μπει ενα exception \n",
    "######## δλδ να ψάχνει αν το γονίδιο που έβαλε ο Χρήστης είνα valid και μετά να κάνει request"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Questio 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1000000']\n",
      "/sequence/region/human/X:1000000..1000100:1?\n",
      "CTGTAGAAACATTAGCCTGGCTAACAAGGTGAAACCCCATCTCTACTAACAATACAAAATATTGGTTGGGCGTGGTGGCGGGTGCTTGTAATCCCAGCTAC\n",
      "101\n"
     ]
    }
   ],
   "source": [
    "import requests, sys\n",
    "import re\n",
    "\n",
    "a=\"What is the sequence in chromosome X between 1000000 and 1000100?\" \n",
    "words = a.split()\n",
    "\n",
    "chrom = [words[i+1] for i,word in enumerate(words) if word.startswith(('chr','Chr', 'cHr','chR', 'CHR', 'cHR', 'ChR'))] \n",
    "digits = [digit for digit in a.split() if digit.isdigit()]\n",
    "print(digits)\n",
    "if len(digits) < 2:\n",
    "    digits.append(words[-1].replace('?', ''))\n",
    "elif len(digits) > 2:\n",
    "    digits=sorted(list(set(digits)-set(chrom)))\n",
    "else:\n",
    "    digits=sorted(digits)\n",
    "\n",
    "server = \"https://rest.ensembl.org\"\n",
    "ext = \"/sequence/region/human/\" + chrom[0] + ':' + str(digits[0]) + '..' + str(digits[1]) + ':1?'\n",
    "print(ext)\n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"text/plain\"})\n",
    " \n",
    "if not r.ok:\n",
    "  r.raise_for_status()\n",
    "  sys.exit()\n",
    " \n",
    " \n",
    "print(r.text)\n",
    "print(len(r.text))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://myvariant.info/v1/variant/chr1:g.218631822G>A\n",
      "According to esp\n",
      "\n",
      "{'allele': 'G', 'freq_1000g': 0.978, 'freq_gnomad': 0.985, 'freq_topmed': 0.983}\n",
      "{'allele': 'A', 'freq_1000g': 0.022, 'freq_gnomad': 0.015, 'freq_topmed': 0.017}\n",
      "Frequency of alleles:\n",
      "('G', 0.978)\n",
      "('A', 0.022)\n",
      "Gnomad Freqs:\n",
      "('G', 0.985)\n",
      "('A', 0.015)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "a=\"What is the allele frequency of <rs58991260> ?\"\n",
    "\n",
    "x = get_the_entry(a)\n",
    "k = mutation(x)\n",
    "\n",
    "from collections import MutableMapping \n",
    "#######https://www.geeksforgeeks.org/python-convert-nested-dictionary-into-flattened-dictionary/\n",
    "def flatten(d, parent_key ='', sep ='_'):\n",
    "\n",
    "    items = [] \n",
    "    for k, v in d.items(): \n",
    "        new_key = parent_key + sep + k if parent_key else k \n",
    "        if isinstance(v, MutableMapping): \n",
    "            items.extend(flatten(v, new_key, sep = sep).items()) \n",
    "        else: \n",
    "            items.append((new_key, v)) \n",
    "    return dict(items)\n",
    "\n",
    "def do_appropriate_mutation_request(k):\n",
    "    \n",
    "    if 'rs' in k:\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': k,\n",
    "            'fields': 'dbsnp'}\n",
    "        d = (requests.get(url, params=parameters)).json()\n",
    "        ID = d['hits'][0]['_id']\n",
    "        url = 'http://myvariant.info/v1/variant/' + ID\n",
    "        print(url)\n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "\n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            \n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "        #except requests.exceptions.Timeout as errt:\n",
    "           # print (\"Timeout Error:\",errt) ## πρέπει να βάλω timeout\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "            \n",
    "        data = (requests.get(url)).json()\n",
    "        try :\n",
    "            dict_freqs = data['cadd']\n",
    "            pass\n",
    "            freqs2 =[ values for keys,values in dict_freqs.items() if keys == 'esp']\n",
    "            first = data['cadd']['1000g']\n",
    "            lists = data['dbsnp']['alleles']\n",
    "\n",
    "            print('According to esp' + '\\n', *freqs2, sep = '\\n')\n",
    "            \n",
    "            freqs = []\n",
    "            gnomad = []\n",
    "            for ds in lists:\n",
    "                dict_mega = flatten(ds)\n",
    "                print(dict_mega)\n",
    "                for k,v in dict_mega.items():\n",
    "                    if k == 'allele' or k == 'freq_1000g':\n",
    "                        freqs.append(v)\n",
    "                    if k == 'allele' or k == 'freq_gnomad':\n",
    "                        gnomad.append(v)\n",
    "            a_freqs = zip(freqs[0::2],freqs[1::2])\n",
    "            gn_freqs = zip(gnomad[0::2],gnomad[1::2])\n",
    "            print(f'Frequency of alleles:',  *a_freqs, sep = '\\n')\n",
    "            print(f'Gnomad Freqs:',  *gn_freqs, sep = '\\n')\n",
    "        \n",
    "        except KeyError:\n",
    "            print(f\"There in no cadd and esp frequencies data for this {k} mutation\")\n",
    "            print(f\"{data['error']}\")\n",
    "            \n",
    "    else:\n",
    "        \n",
    "        url = 'http://myvariant.info/v1/variant/' + k\n",
    "        \n",
    "        print(url)\n",
    "        \n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "\n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "            \n",
    "        data = (requests.get(url)).json()\n",
    "        \n",
    "        try:\n",
    "            first = data['cadd']['1000g']\n",
    "            pass\n",
    "            af =  data['cadd']['1000g']['af']\n",
    "            afr = data['cadd']['1000g']['afr']\n",
    "            amr =  data['cadd']['1000g']['amr']\n",
    "            eur =  data['cadd']['esp']['eur']\n",
    "            lists = data['dbsnp']['alleles']\n",
    "            print('According to 1000g' + '\\n' + f'af: {af}, afr: {afr}, amr: {amr}, eur: {eur}')\n",
    "            \n",
    "            freqs = []\n",
    "            gnomad = []\n",
    "            for ds in lists:\n",
    "                dict_mega = flatten(ds)\n",
    "                print(dict_mega)\n",
    "                for k,v in dict_mega.items():\n",
    "                    if k == 'allele' or k == 'freq_1000g':\n",
    "                        freqs.append(v)\n",
    "                    if k == 'allele' or k == 'freq_gnomad':\n",
    "                        gnomad.append(v)\n",
    "            a_freqs = zip(freqs[0::2],freqs[1::2])\n",
    "            gn_freqs = zip(gnomad[0::2],gnomad[1::2])\n",
    "            print(f'Frequency of alleles:',  *a_freqs, sep = '\\n')\n",
    "            print(f'Gnomad Freqs:',  *gn_freqs, sep = '\\n')\n",
    "        except KeyError:\n",
    "            print(f\"There in no cadd and 1000q frequencies data for this {k} mutation\")\n",
    "            print(f\"{data['error']}\")\n",
    "   \n",
    "        \n",
    "do_appropriate_mutation_request(k)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "According to esp\n",
      "\n",
      "{'af': 0.0, 'afr': 0.001, 'eur': 0.0}\n"
     ]
    }
   ],
   "source": [
    "from collections import MutableMapping \n",
    "#######https://www.geeksforgeeks.org/python-convert-nested-dictionary-into-flattened-dictionary/\n",
    "def flatten(d):\n",
    "    parent_key =''\n",
    "    sep ='_'\n",
    "    items = [] \n",
    "    for k, v in d.items(): \n",
    "        new_key = parent_key + sep + k if parent_key else k \n",
    "        if isinstance(v, MutableMapping): \n",
    "            items.extend(convert_flatten(v, new_key, sep = sep).items()) \n",
    "        else: \n",
    "            items.append((new_key, v)) \n",
    "    return dict(items)\n",
    "\n",
    "url = 'http://myvariant.info/v1/variant/chr9:g.107620835G>A'\n",
    "data = requests.get(url)\n",
    "d=data.json()\n",
    "dict_freqs = d['cadd']\n",
    "freqs2 =[ values for keys,values in dict_freqs.items() if keys == 'esp']\n",
    "print('According to esp' + '\\n', *freqs2, sep = '\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'get_the_entry' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-b77d735f891e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_the_entry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0mk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmutation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'get_the_entry' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "a='With which conditions is the mutation <chr9:g.107620835G>A> associated ?'\n",
    "\n",
    "\n",
    "x = get_the_entry(a)\n",
    "k = mutation(x)\n",
    "\n",
    "def do_appropriate_mutation_request(k):\n",
    "    \n",
    "    if 'rs' in k:\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': k,\n",
    "            'fields': 'dbsnp'}\n",
    "        d = (requests.get(url, params=parameters)).json()\n",
    "        ID = d['hits'][0]['_id']\n",
    "        url = 'http://myvariant.info/v1/variant/' + ID\n",
    "        print(url)\n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "\n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            \n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "        #except requests.exceptions.Timeout as errt:\n",
    "           # print (\"Timeout Error:\",errt) ## πρέπει να βάλω timeout\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "            \n",
    "        data = (requests.get(url)).json()\n",
    "        try :\n",
    "            cond_1 = data['clinvar']['rcv']\n",
    "            pass\n",
    "            cond_2 = [v.get('name') for d in cond_1 for k,v in d.items() if  isinstance(v,dict)]\n",
    "            print(f'According to clinvar:' + '\\n', *cond_2, sep='\\n')\n",
    "        except KeyError:\n",
    "            print(f\"There in no clinvar data for this {k} mutation\")\n",
    "            print(f\"{data['error']}\")\n",
    "   \n",
    "        \n",
    "    else:\n",
    "        \n",
    "        url = 'http://myvariant.info/v1/variant/' + k\n",
    "        \n",
    "        print(url)\n",
    "        \n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "\n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            #print(f\"{data['error']}\")\n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "        #except requests.exceptions.Timeout as errt:\n",
    "           # print (\"Timeout Error:\",errt) ## πρέπει να βάλω timeout\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "            \n",
    "        data = (requests.get(url)).json()\n",
    "        \n",
    "        try:\n",
    "            cond = data['clinvar']['rcv']  \n",
    "            pass\n",
    "            cond2 = [v.get('name') for d in cond for k,v in d.items() if  isinstance(v,dict)]\n",
    "            print(f'According to clinvar:' + '\\n', *cond2, sep='\\n')\n",
    "        except KeyError:\n",
    "            print(f\"There in no clinvar data for this {k} mutation\")\n",
    "            print(f\"{data['error']}\")\n",
    "   \n",
    "            \n",
    "\n",
    "do_appropriate_mutation_request(k)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr9:g.107620835G>A\n",
      "http://myvariant.info/v1/variant/chr9:g.107620835G>A\n",
      "According to clinvar:\n",
      "\n",
      "Tangier disease (TGD)\n",
      "Familial High Density Lipoprotein Deficiency\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "########chr9:g.107620835G>A\n",
    "## works rs9282541\n",
    "## do not rs12345678\n",
    "a='With which conditions is the mutation <chr9:g.107620835G>A> associated ?'\n",
    "\n",
    "def get_the_entry(y):\n",
    "    mutation = (re.search(r\"\\<([A-Za-z0-9_.:>]+)\\>\", y).group(1))\n",
    "    return mutation\n",
    "\n",
    "def get_chrom_transcipt(x): \n",
    "    \n",
    "    url = 'http://mygene.info/v3/query'\n",
    "    parameters = {\n",
    "      'fields': 'genomic_pos',\n",
    "      'species': 'human',\n",
    "      'q' : 'ensembl.transcript:' + x}\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data = response.json()\n",
    "    return data['hits'][0]['genomic_pos']['chr']\n",
    "\n",
    "def get_chrom_gene(x): \n",
    "    url = 'http://myvariant.info/v1/query'\n",
    "    parameters = {\n",
    "        'q': x,\n",
    "        'fields': 'dbsnp'}\n",
    "    d = (requests.get(url, params=parameters)).json()\n",
    "    chromosome = d['hits'][0]['dbsnp']['chrom']\n",
    "    return chromosome\n",
    "\n",
    "\n",
    "def mutation(x):\n",
    "    \n",
    "        if 'rs' in x:\n",
    "            return x\n",
    "        elif x.startswith(('chr','Chr', 'cHr','chR', 'CHR', 'cHR', 'ChR')):\n",
    "            if '>' in x:\n",
    "                return x\n",
    "        elif x[0].isdigit():\n",
    "                return 'chr'+x\n",
    "        elif 'c.' in x:\n",
    "            if x.startswith(('ENST0', 'NM')):\n",
    "                position = x.split('.')[1]\n",
    "                transcript = re.search(r\"([A-Za-z0-9_]+)\", a).group()\n",
    "                mutation = 'chr' + get_chrom_transcipt(transcript) + ':' + position\n",
    "                return  mutation\n",
    "            else:\n",
    "                position = x.split('.')[1]\n",
    "                gene = re.search(r\"([A-Za-z0-9_]+)\", a).group()\n",
    "                mutation='chr'+ get_chrom_gene(gene) + ':' + position\n",
    "                return  mutation\n",
    "        else:\n",
    "            lista_formats=['rs12345678', 'chr1:1234567A>G', '1:1234567A>G', '<Gene>:c.100A>G', \n",
    "                           '<Transcript (Ensembl or Refeseq_ID)>:c.100A>G']\n",
    "            raise Exception(\"Please specify your mutation properly. The examples for acceptable formats are:\" +'\\n'+ \n",
    "                            f\"{lista_formats}\")\n",
    "            \n",
    "k = mutation(get_the_entry(a))\n",
    "print(k)\n",
    "\n",
    "def do_appropriate_mutation_request(k):\n",
    "    \n",
    "    if 'rs' in k:\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': k,\n",
    "            'fields': 'dbsnp'}\n",
    "        d = (requests.get(url, params=parameters)).json()\n",
    "        ID = d['hits'][0]['_id']\n",
    "        url = 'http://myvariant.info/v1/variant/' + ID\n",
    "        print(url)\n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "\n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            \n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "        #except requests.exceptions.Timeout as errt:\n",
    "           # print (\"Timeout Error:\",errt) ## πρέπει να βάλω timeout\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "            \n",
    "        data = (requests.get(url)).json()\n",
    "        try :\n",
    "            cond_1 = data['clinvar']['rcv']\n",
    "            pass\n",
    "            cond_2 = [v.get('name') for d in cond_1 for k,v in d.items() if  isinstance(v,dict)]\n",
    "            print(f'According to clinvar:' + '\\n', *cond_2, sep='\\n')\n",
    "        except KeyError:\n",
    "            print(f\"There in no clinvar data for this {k} mutation\")\n",
    "            print(f\"{data['error']}\")\n",
    "   \n",
    "        \n",
    "    else:\n",
    "        \n",
    "        url = 'http://myvariant.info/v1/variant/' + k\n",
    "        \n",
    "        print(url)\n",
    "        \n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "\n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            #print(f\"{data['error']}\")\n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "        #except requests.exceptions.Timeout as errt:\n",
    "           # print (\"Timeout Error:\",errt) ## πρέπει να βάλω timeout\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "            \n",
    "        data = (requests.get(url)).json()\n",
    "        \n",
    "        try:\n",
    "            cond = data['clinvar']['rcv']  \n",
    "            pass\n",
    "            cond2 = [v.get('name') for d in cond for k,v in d.items() if  isinstance(v,dict)]\n",
    "            print(f'According to clinvar:' + '\\n', *cond2, sep='\\n')\n",
    "        except KeyError:\n",
    "            print(f\"There in no clinvar data for this {k} mutation\")\n",
    "            print(f\"{data['error']}\")\n",
    "   \n",
    "            \n",
    "\n",
    "do_appropriate_mutation_request(k)\n",
    "\n",
    "############### πρέπει να μπει ένα  exception όταν δεν μπει κάποιο σωστό mutation; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr9:g.107620835G>A\n",
      "http://myvariant.info/v1/variant/chr9:g.107620835G>A\n",
      "The rsid of mutation chr9:g.107620835G>A is: rs9282541\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import re\n",
    "\n",
    "#a='ENST00000003084'\n",
    "#a='chr9:g.107620835G>A'\n",
    "a='With which conditions is the mutation <chr9:g.107620835G>A> associated ?'\n",
    "\n",
    "#a='TPMT'\n",
    "\n",
    "def get_the_entry(y):\n",
    "    mutation = (re.search(r\"\\<([A-Za-z0-9_.:>]+)\\>\", y).group(1))\n",
    "    return mutation\n",
    "\n",
    "def get_chrom_transcipt(x): \n",
    "    parameters = {\n",
    "      'fields': 'genomic_pos',\n",
    "      'species': 'human',\n",
    "      'q' : 'ensembl.transcript:' + x,\n",
    "                        }\n",
    "    url = 'http://mygene.info/v3/query'\n",
    "\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data = response.json()\n",
    "    return data['hits'][0]['genomic_pos']['chr']\n",
    "\n",
    "#get_chrom_transcipt(a)\n",
    "\n",
    "\n",
    "def get_chrom_gene(x): \n",
    "    url = 'http://myvariant.info/v1/query'\n",
    "    parameters = {\n",
    "        'q': x,\n",
    "        'fields': 'dbsnp',\n",
    "    }\n",
    "\n",
    "    d = (requests.get(url, params=parameters)).json()\n",
    "    chromosome = d['hits'][0]['dbsnp']['chrom']\n",
    "    return chromosome\n",
    "#get_chrom_gene(a)\n",
    "\n",
    "\n",
    "def mutation(x):\n",
    "    \n",
    "        if 'rs' in x:\n",
    "            return x\n",
    "        elif x.startswith(('chr','Chr', 'cHr','chR', 'CHR', 'cHR', 'ChR')):\n",
    "            if '>' in x:\n",
    "                return x\n",
    "        elif x[0].isdigit():\n",
    "                return 'chr'+x\n",
    "        elif 'c.' in x:\n",
    "            if x.startswith(('ENST0', 'NM')):\n",
    "                position = x.split('.')[1]\n",
    "                transcript = re.search(r\"([A-Za-z0-9_]+)\", a).group()\n",
    "                mutation = 'chr' + get_chrom_transcipt(transcript) + ':' + position\n",
    "                return  mutation\n",
    "            else:\n",
    "                position = x.split('.')[1]\n",
    "                gene = re.search(r\"([A-Za-z0-9_]+)\", a).group()\n",
    "                mutation='chr'+ get_chrom_gene(gene) + ':' + position\n",
    "                return  mutation\n",
    "#mutation(a)\n",
    "\n",
    "mutation(get_the_entry(a))\n",
    "k = mutation(get_the_entry(a))\n",
    "print(k)\n",
    "\n",
    "def do_appropriate_mutation_request(k):\n",
    "    \n",
    "    if 'rs' in k:\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "        'q:'+ k,\n",
    "        'fields:' 'dbsnp.rsid'}\n",
    "        \n",
    "        data = (requests.get(url, params=parameters)).json()\n",
    "        #print(data)\n",
    "        rsid = data['hits'][0]['dbsnp']['rsid']\n",
    "        print(url)\n",
    "        print(rsid)\n",
    "        print(f'The rsid of mutation {k} is: {rsid}')\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        url = 'http://myvariant.info/v1/variant/' + k\n",
    "        data = (requests.get(url)).json()\n",
    "        print(url)\n",
    "        rsid = data['clinvar']['rsid']\n",
    "        print(f'The rsid of mutation {k} is: {rsid}')\n",
    "        \n",
    "do_appropriate_mutation_request(k)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 20\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr6:g.18149300T>A\n",
      "http://myvariant.info/v1/variant/chr6:g.18149300T>A\n",
      "According to:\n",
      "snpeff: missense_variant\n",
      "Polyphen: 0.144(benign)\n",
      "SIFT: converted_rankscore:0.51248, pred:'D', score: 0.017\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "########chr9:g.107620835G>A\n",
    "## works rs9282541\n",
    "## do not rs12345678\n",
    "a='With which conditions is the mutation <chr6:g.18149300T>A> associated ?'\n",
    "\n",
    "def get_the_entry(y):\n",
    "    mutation = (re.search(r\"\\<([A-Za-z0-9_.:>]+)\\>\", y).group(1))\n",
    "    return mutation\n",
    "\n",
    "def get_chrom_transcipt(x): \n",
    "    \n",
    "    url = 'http://mygene.info/v3/query'\n",
    "    parameters = {\n",
    "      'fields': 'genomic_pos',\n",
    "      'species': 'human',\n",
    "      'q' : 'ensembl.transcript:' + x}\n",
    "    response = requests.get(url, params=parameters)\n",
    "    data = response.json()\n",
    "    return data['hits'][0]['genomic_pos']['chr']\n",
    "\n",
    "def get_chrom_gene(x): \n",
    "    url = 'http://myvariant.info/v1/query'\n",
    "    parameters = {\n",
    "        'q': x,\n",
    "        'fields': 'dbsnp'}\n",
    "    \n",
    "    d = (requests.get(url, params=parameters)).json()\n",
    "    chromosome = d['hits'][0]['dbsnp']['chrom']\n",
    "    return chromosome\n",
    "\n",
    "\n",
    "def mutation(x):\n",
    "    \n",
    "        if 'rs' in x:\n",
    "            return x\n",
    "        elif x.startswith(('chr','Chr', 'cHr','chR', 'CHR', 'cHR', 'ChR')):\n",
    "            if '>' in x:\n",
    "                return x\n",
    "        elif x[0].isdigit():\n",
    "                return 'chr'+x\n",
    "        elif 'c.' in x:\n",
    "            if x.startswith(('ENST0', 'NM')):\n",
    "                position = x.split('.')[1]\n",
    "                transcript = re.search(r\"([A-Za-z0-9_]+)\", a).group()\n",
    "                mutation = 'chr' + get_chrom_transcipt(transcript) + ':' + position\n",
    "                return  mutation\n",
    "            else:\n",
    "                position = x.split('.')[1]\n",
    "                gene = re.search(r\"([A-Za-z0-9_]+)\", a).group()\n",
    "                mutation='chr'+ get_chrom_gene(gene) + ':' + position\n",
    "                return  mutation\n",
    "        else:\n",
    "            lista_formats=['rs12345678', 'chr1:1234567A>G', '1:1234567A>G', '<Gene>:c.100A>G', \n",
    "                           '<Transcript (Ensembl or Refeseq_ID)>:c.100A>G']\n",
    "            raise Exception(\"Please specify your mutation properly. The examples for acceptable formats are:\" +'\\n'+ \n",
    "                            f\"{lista_formats}\")\n",
    "            \n",
    "k = mutation(get_the_entry(a))\n",
    "print(k)\n",
    "\n",
    "def do_appropriate_mutation_request(k):\n",
    "    \n",
    "    if 'rs' in k:\n",
    "        url = 'http://myvariant.info/v1/query'\n",
    "        parameters = {\n",
    "            'q': k,\n",
    "            'fields': 'dbsnp'}\n",
    "        d = (requests.get(url, params=parameters)).json()\n",
    "        ID = d['hits'][0]['_id']\n",
    "        url = 'http://myvariant.info/v1/variant/' + ID\n",
    "        print(url)\n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()\n",
    "            \n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            #print(f\"{data['error']}\")\n",
    "            \n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "            \n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "         \n",
    "        \n",
    "        data = (requests.get(url)).json()\n",
    "        \n",
    "        try :\n",
    "            effect = data['snpeff']['ann']['effect']\n",
    "            print('According to:' + '\\n' + f\"snpeff: {effect}\")\n",
    "        except KeyError:\n",
    "            print(f\"There in no snpeff data for effect of this {k} mutation\")\n",
    "                  \n",
    "        try:\n",
    "            poly_val = data['cadd']['polyphen']['val']\n",
    "            pass\n",
    "            poly_cat = data['cadd']['polyphen']['cat']\n",
    "            print(f\"Polyphen: {poly_val}({poly_cat})\")\n",
    "                  \n",
    "        except KeyError:\n",
    "            print(f\"There in no cadd data for polyphen of this {k} mutation\")\n",
    "        \n",
    "        try:\n",
    "            cr = data['dbnsfp']['sift']['converted_rankscore']\n",
    "            pass    \n",
    "            pred = data['dbnsfp']['sift']['pred']\n",
    "            score = data['dbnsfp']['sift']['score']\n",
    "            print(f\"SIFT: converted_rankscore:{cr}, pred:'{pred}', score: {score}\")\n",
    "                  \n",
    "        except KeyError:\n",
    "            print(f\"There in dbsnf data for this {k} mutation\")\n",
    "   \n",
    "        \n",
    "    else:\n",
    "        \n",
    "        url = 'http://myvariant.info/v1/variant/' + k\n",
    "        \n",
    "        print(url)\n",
    "        \n",
    "        try: \n",
    "            r = requests.get(url)\n",
    "            r.raise_for_status()   \n",
    "                  \n",
    "        except requests.exceptions.HTTPError as errh:\n",
    "            print (\"Http Error:\",errh)\n",
    "            #print(f\"{data['error']}\")\n",
    "                  \n",
    "        except requests.exceptions.ConnectionError as errc:\n",
    "            print (\"Error Connecting:\",errc)\n",
    "\n",
    "        except requests.exceptions.RequestException as err:\n",
    "            print (\"OOps: Something Else\",err)\n",
    "                  \n",
    "            effect = data['snpeff']['ann']['effect']\n",
    "            print('According to:' + '\\n' + f\"snpeff: {effect}\")     \n",
    "        except KeyError:\n",
    "            print(f\"There in no snpeff data for effect of this {k} mutation\")\n",
    "                  \n",
    "        data = (requests.get(url)).json()\n",
    "        \n",
    "        try :\n",
    "            effect = data['snpeff']['ann']['effect']\n",
    "            print('According to:' + '\\n' + f\"snpeff: {effect}\")\n",
    "        except KeyError:\n",
    "            print(f\"There in no snpeff data for effect of this {k} mutation\")\n",
    "                  \n",
    "        try:\n",
    "            poly_val = data['cadd']['polyphen']['val']\n",
    "            pass\n",
    "            poly_cat = data['cadd']['polyphen']['cat']\n",
    "            print(f\"Polyphen: {poly_val}({poly_cat})\")\n",
    "                  \n",
    "        except KeyError:\n",
    "            print(f\"There in no cadd data for polyphen of this {k} mutation\")\n",
    "        \n",
    "        try:\n",
    "            cr = data['dbnsfp']['sift']['converted_rankscore']\n",
    "            pass    \n",
    "            pred = data['dbnsfp']['sift']['pred']\n",
    "            score = data['dbnsfp']['sift']['score']\n",
    "            print(f\"SIFT: converted_rankscore:{cr}, pred:'{pred}', score: {score}\")\n",
    "                  \n",
    "        except KeyError:\n",
    "            print(f\"There in dbsnf data for this {k} mutation\")\n",
    "            \n",
    "do_appropriate_mutation_request(k)\n",
    "\n",
    "############### πρέπει να μπει ένα  exception όταν δεν μπει κάποιο σωστό mutation; done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
